optuna/samplers/_tpe/sampler.py

```diff
@@ -742,52 +750,30 @@ def _calculate_weights_below_for_multi_objective(
-        if constraints_func is not None:
-            if any(constraint > 0 for constraint in constraints_func(trial)):
-                feasible_mask[i] = False
-                continue
-        values = []
-        for value, direction in zip(trial.values, study.directions):
-            if direction == StudyDirection.MINIMIZE:
-                values.append(value)
-            else:
-                values.append(-value)
-        loss_vals.append(values)
-    lvals = np.asarray(loss_vals, dtype=float)
-
-    # Calculate weights based on hypervolume contributions.
-    n_below = len(lvals)
-    weights_below: np.ndarray
-    if n_below == 0:
-        weights_below = np.asarray([])
-    elif n_below == 1:
-        weights_below = np.asarray([1.0])
-    else:
-        worst_point = np.max(lvals, axis=0)
-        reference_point = np.maximum(1.1 * worst_point, 0.9 * worst_point)
-        reference_point[reference_point == 0] = EPS
-        hv = compute_hypervolume(lvals, reference_point)
-        indices_mat = ~np.eye(n_below).astype(bool)
-        contributions = np.asarray(
-            [
-                hv - compute_hypervolume(lvals[indices_mat[i]], reference_point)
-                for i in range(n_below)
-            ]
-        )
-        contributions[np.isnan(contributions)] = np.inf
-        max_contribution = np.maximum(np.max(contributions), EPS)
-        if not np.isfinite(max_contribution):
-            weights_below = np.ones_like(contributions, dtype=float)
-            # TODO(nabenabe0928): Make the weights for non Pareto solutions to zero.
-            weights_below[np.isfinite(contributions)] = EPS
-        else:
-            weights_below = np.clip(contributions / max_contribution, EPS, 1)
+    def _feasible(trial: FrozenTrial) -> bool:
+        return constraints_func is None or all(c <= 0 for c in constraints_func(trial))
+
+    is_feasible = np.asarray([_feasible(t) for t in below_trials])
+    weights_below = np.where(is_feasible, 1.0, EPS)  # Assign EPS to infeasible trials.
+    n_below_feasible = np.count_nonzero(is_feasible)
+    if n_below_feasible <= 1:
+        return weights_below
 
-    # For now, EPS weight is assigned to infeasible trials.
-    weights_below_all = np.full(len(below_trials), EPS)
-    weights_below_all[feasible_mask] = weights_below
-    return weights_below_all
+    lvals = np.asarray([t.values for t in below_trials])[is_feasible]
+    lvals *= np.array([-1.0 if d == StudyDirection.MAXIMIZE else 1.0 for d in study.directions])
+    ref_point = _get_reference_point(lvals)
+    on_front = _is_pareto_front(lvals, assume_unique_lexsorted=False)
+    pareto_sols = lvals[on_front]
+    hv = compute_hypervolume(pareto_sols, ref_point, assume_pareto=True)
+    if np.isinf(hv):
+        # TODO(nabenabe): Assign EPS to non-Pareto solutions, and
+        # solutions with finite contrib if hv is inf. Ref: PR#5813.
+        return weights_below
```

Comment:
Maybe we should, first of all, replace $\pm \infty$ with a big value like here:

- https://github.com/optuna/optuna/pull/5798/files#diff-152bf351e020a2e950295910a479597f1026e376322ef4c934e4f04ad6f69fafR119-R124